# Analysis of antihelminthic treatment dataset

 [![License CCO v.1](https://img.shields.io/github/license/PollyTikhonova/Taye_AntihelminthicTreatment_2025?v=1
)](https://github.com/PollyTikhonova/Taye_AntihelminthicTreatment_2025/blob/main/LICENSE)
 doi shield

Code accompanying the manuscript "Microbiome responses to anthelmintic treatment are contingent on pre-treatment helminth infection status in young Ethiopian children: A longitudinal study".

Abstract: Mass deworming programs using effective anthelmintic drugs are essential for controlling soil-transmitted helminth (STH) infections, particularly in high-risk developing regions. However, it remains unclear whether routine deworming induces long-term alterations in gut microbiome composition, especially when accounting for individual infection status. This study aims to explore the changes in the gut microbiomes of Ethiopian children one year after the administration of an anthelmintic treatment.

### Contributors
- Polina Tikhonova, MSc


### Overview

	project
	|- README         # repository descriptions
  	|- LICENSE        # the license for this project
	|
	|- data           # project data
	| |- database/    # SILVA database directory
	| |- reads/       # a directory for fastq files 
  	| |- raw/      # a directory for the raw reads
   	| +- filtered/ # (generated by the pipeline)
   	|                 # a directory, containing filtered reads
   	| |- QCcontrol    # (generated by the pipeline); QC files
   	| |- dada2        # (generated by the pipeline); dada2 files
   	| |- phyloseq     # (generated by the pipeline; provided);
   	| |               # resulting phyloseq object files in .rds and .csv formats
   	| +- metadata.tsv # a metadata file, provided
	|
	|- code/          # project codes
	| |- analysis/    # analysis, generates results/ files
	| +- processing/  # raw data processing, generates data/proccessed/ files
	|
	|- results        # output files, generated by the codes in analysis folder
	| |- figures/     # manuscript figures
	| +- tables/      # beta-diversity tables


## Code descriptions

### Data Processing
#### Dependencies
* snakemake (v7.32.1) <br>
```conda install -c bioconda -c conda-forge snakemake==7.32.1```

#### Instructions
_Please note that running the processing pipeline is optional since the final phyloseq objects necessary for the analysis are provided in this repository._
1. Install dependencies.
2. The raw data is publicly available at the European Nucleotide Archive (ENA) under accession number [PRJEB93790](https://www.ebi.ac.uk/ena/browser/view/PRJEB93790). Please download the fastq.gz files to the `data/reads/raw` directory and follow the proposed directory structure: `sample_name/sample_name_R*.fastq.qz`. The metadata file should be saved as a `data/metadata.tsv` file (a tab-separated format). 
3. Download Silva database version 132 (https://zenodo.org/records/1172783) to the `data/database` directory.
4. (optional) In case of custom paths to the raw files and database, please modify the corresponding parameters in the `code/processing/config.yaml` file.
5. Run Snakemodule
``` 
mkdir code/processing/logs
cd code/processing/logs
conda activate snakemake
snakemake --snakefile ../Snakemodule --use-conda --conda-frontend conda
```
Suggested parameters for running the Snakemodule using multiple sbatch jobs: `--cluster "sbatch --time=47:00:00 --nodes=1 --ntasks=20 --mem=200GB" --jobs 10`. 

#### Pipeline files
```
codes/processing
|
|- Snakemodule
|- conda_env.*.yaml                      # conda environment libraries
|- config.yaml                           # Snakemodule paths and settings
+- params.dada2.filter_and_trim.yaml     # dada2 filter_and_trim settings
```

__Snakemodule Steps__
1. FastQC/MultiQC quality assessment
   - `1.quality_control.smk`
2. Trimming and filtering
   - `2.filtering_trimming.smk`
3. FastQC/MultiQC quality reassessment
   - `1.quality_control.smk`
4. ASV identification using DADA2 pipeline [(official tutorial)](https://benjjneb.github.io/dada2/tutorial.html)
   - `3.dada2.smk`
5. ASV taxonomy annotation
   - `3.dada2.smk`
6. Generation of the phyloseq object.
   - `3.dada2.smk`
   - outputs: 
     - unfiltered phyloseq (all ASVs)
     - filtered phyloseq (only microbial ASVs)
     - filtered phyloseq objects agglomerated at genus and family levels

---
### Analysis
#### Dependencies
1. Create and install a new conda environment
```
conda create -n anthelminthic_treatment
conda env update -n anthelminthic_treatment --file code/analysis/conda_env.R.yaml
```
2. Install additional R libraries:
```
conda activate anthelminthic_treatment
R
devtools::install_github(repo = "malucalle/selbal")
devtools::install_github(repo = "gauravsk/ranacapa")
devtools::install_github(repo = "gmteunisse/fantaxtic")
```

#### Files
All analysis codes are implemented in R and stored in the `codes/processing` directory. Each coding file is provided in two extensions: `.ipynb` and `.Rmd`
1. Dataset overview
   - `Figure 1. Data Overview.` (_Chi-squared test of demographic characteristics, sample counts_).
   - `Figure 3. Microbial relative abundance barplots.`
2. Microbial diversity
   - `Figure 2. Microbial diversity.` (_baseline vs follow-up groups_).
   - `Figure 2. Microbial diversity. Accounting for STH status.` (_baseline vs follow-up groups_).
3. Dimensionality reduction
   - `Figure 2. PCoA based on Bray-Curtis distance. Baseline and follow-up groups.`
   - `Supplementary Figure 5. PCoA based on Bray-Curtis distance. Prior antihelminthic treatment.`
4. Genera-vanishment analysis
   - `Figure 3. ASV vanishment analysis`
5. Beta-diversity
   - `Figure 4. ALDEx2 analysis. Wilcoxon test.`
   - `Supplementary Figure 4. ALDEx2 analysis. Welch test.`
   - `Supplementary Figure 5. DESeq2 analysis. `
6. Prediction of previous antihelminthic treatments using microbial balance score
   - `Figure 5. Prediction of previous antihelminthic treatments.`
